Add docker-compose testnet

mixmasala opened the pull request:
This adds Makefile targets to start and stop a local talek deployment using docker-compose and podman or docker. It provides a docker-compose.yml and writes a client configuration file for use with talekclient.

mixmasala committed:
update Dockerfile to golang:1.20 and go mod tidy

mixmasala committed:
Makefile, Dockerfile: add talek testnet make targets

This adds Makefile targets for running a local deployment of talek with podman and docker-compose.

testnet-start: starts a talek deployment
testnet-stop: stops the deployment
testnet-build-config: builds configuration with talekutil
testnet-clean: cleans images and stampfiles
testnet-cli: starts a shell with talek cli programs

mixmasala commented:
You should be able to replicate this by using the Makefile target testnet-start and testnet-cli, where the client configuration is written to /talek_net/talek.json inside the container

willscott commented on talek_net/docker-compose.yml:
```diff
@@ -0,0 +1,39 @@
+    restart: "no"
+    image: talek
+    volumes:
+      - ./:/talek_net
```
maybe make the volume `talek_shared_mnt` or something that disambiguates from the docker network `net`?

mixmasala replied:
I changed the volume path to talek_shared

mixmasala committed:
make volume mount path talek_shared, use talek-cli image

mixmasala committed:
fix 2 volume mount paths, mount to talek-shared

mixmasala committed:
set FrontendAddr in client configuration

mixmasala committed:
index replicas from 0

mixmasala committed:
use correct replica configuration names

mixmasala committed:
talekclient: parse read-only share

fixes #108

leif committed:
Merge branch 'master' into add_102_talek_testnet

leif committed:
docker testnet: use sed to edit json config

instead of requiring a python docker image

leif committed:
Makefile: robustify cleanup

leif committed:
docker testnet: cleanup

- use default config file names
- use docker workdir
- escape quotes so sed line actually works
- fix port number

leif committed:
Makefile: add testnet-test-write-and-read target

This works for me right now, but I don't know why I need to fork the log handle
into a reader and a writer the way I am here before using it.

If I read and write using the same handle (without making a copy of it first),
the read operations always read from the locations --write would write to
*next*, rather than the previous locations which have actually been written to.
So, there is either an off-by-one error somewhere or I am confused about how
this is supposed to work.

leif committed:
Makefile: cleanup

- allow stopping and restarting without regenerating configs
- make testnet-test-write-and-read target depend on the network being up

leif committed:
Merge branch 'fix_108_talekclient_parse_share'

leif committed:
testnet-test-write-and-read: use --share

After merging masala's fix for parsing read-only handles from
https://github.com/privacylab/talek/pull/110 this works now.

willscott commented:
Is this a a point that's ready for a review?

leif commented:
> Is this a a point that's ready for a review?

Yes please. Note that this has the fix from the other open PR (#110) merged to make the write-and-read integration test work.

Let me know if you're actually running travis somewhere and I could add it to `.travis.yml` too (or if you want to setup github CI to run it i could do that also/instead).

willscott commented:
* I am not aware of a travis instance running somewhere
* happy to get github CI set up as well
